### DISCUSSION TITLE
How to manage a transaction for session from async_scoped_session?

### DISCUSSION THREAD

**Discussion by albertalexandrov** [sqlalchemy/sqlalchemy#11152] [Usage Questions]
Hi!

**Original question with FastAPI surroundings**: [https://stackoverflow.com/questions/78133076/override-bind-for-existing-session-to-make-it-join-external-transaction](https://stackoverflow.com/questions/78133076/override-bind-for-existing-session-to-make-it-join-external-transaction) 

I'm trying to create a decorator that could manage a transaction (open/close/commit/rollback). But the difficulty is that session already create by calling `async_scoped_session`. Please look at the code below:

```py
import random
from asyncio import current_task
from functools import wraps

from pydantic import BaseModel
from sqlalchemy import MetaData
from sqlalchemy.ext.asyncio import async_scoped_session, async_sessionmaker, create_async_engine, AsyncSession
from sqlalchemy.orm import declarative_base, Mapped, mapped_column

engine = create_async_engine("url", echo=True)

scoped_session = async_scoped_session(
    async_sessionmaker(bind=engine, expire_on_commit=False, autocommit=False, autoflush=False),
    scopefunc=current_task
)

metadata = MetaData()
Base = declarative_base(metadata=metadata)


class Country(Base):
    id: Mapped[int] = mapped_column(autoincrement=True, primary_key=True, index=True)
    code: Mapped[str]
    name: Mapped[str]


def atomic(func):
    @wraps(func)
    async def wrapper(*args, **kwargs):
        # here I want to manage an external transaction so that only commit in here committed changes to database
        # I mean commits inside wrapped function wouldn't make commit and could be rollbacked here
        async with engine.begin() as connection:
            session = scoped_session()
            # session already in async_scoped_session.registry.registry
            session.bind = connection  # this does not work
            result = await func(*args, **kwargs)

        return result

    return wrapper


class TestTransactionDataSchema(BaseModel):
    first_name: str
    last_name: str


@atomic
async def test_external_transaction():
    session = scoped_session()
    country = Country(code=str(random.randint(1, 100_000)), name=str(random.randint(1, 100_000)))
    session.add(country)
    await session.commit()
    country = Country(code=str(random.randint(1, 100_000)), name=str(random.randint(1, 100_000)))
    session.add(country)
    await session.commit()


def init():
    # because this is simplified version of real code (with FastAPI)
    # I have to init session here to put session into async_scoped_session registry to simulate real situation
    # (you can see real surroundings on stackoverflow question https://stackoverflow.com/questions/78133076/override-bind-for-existing-session-to-make-it-join-external-transaction)
    scoped_session()
    

init()
```

How could I manage transaction inside decorator? How could I make it rollback commits inside `test_external_transaction` function?

**Comment by CaselIT** [comment]
Hi,

Have you tried temove `bind=engine` from `async_sessionmaker` and pass it to when calling the scoped session:
```py
        async with engine.begin() as connection:
            session = scoped_session(connection)
```

Also you may want to remove the session once you are done with them. See https://docs.sqlalchemy.org/en/20/orm/contextual.html#sqlalchemy.orm.scoped_session.remove (same apply in the async case)

**Reply by albertalexandrov** [reply]
Yeah. Because session already instantiated I got the error (in `atomic` decorator):

`sqlalchemy.exc.InvalidRequestError: Scoped session is already present; no new arguments may be specified.`

**Reply by CaselIT** [reply]
I don't understand who has already instanciated the session then.
Can't you ensure that the atomic decorator does the intanciation?

**Reply by albertalexandrov** [reply]
FastAPI instantiates. The question https://stackoverflow.com/questions/78133076/override-bind-for-existing-session-to-make-it-join-external-transaction

**Reply by CaselIT** [reply]
As far as I know it's not possible. Don't make fast api instanciate it, or create a transaction directly using the session

**Reply by albertalexandrov** [reply]
> create a transaction directly using the session

Could you please explain?

**Reply by CaselIT** [reply]
well session has the ability of creating savepoints etc: https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.begin

**Reply by albertalexandrov** [reply]
I tried this (just `begin` has no effect):

```
def atomic(func):

    @wraps(func)
    async def wrapper(*args, **kwargs):
        session = scoped_session()
        await session.begin_nested()
        result = await func(*args, **kwargs)
        return result

    return wrapper
```

Logs:

> 2024-03-13 11:52:45,106 INFO sqlalchemy.engine.Engine select pg_catalog.version()
> 2024-03-13 11:52:45,106 INFO sqlalchemy.engine.Engine [raw sql] ()
> 2024-03-13 11:52:45,112 INFO sqlalchemy.engine.Engine select current_schema()
> 2024-03-13 11:52:45,112 INFO sqlalchemy.engine.Engine [raw sql] ()
> 2024-03-13 11:52:45,123 INFO sqlalchemy.engine.Engine show standard_conforming_strings
> 2024-03-13 11:52:45,123 INFO sqlalchemy.engine.Engine [raw sql] ()
> 2024-03-13 11:52:45,133 INFO sqlalchemy.engine.Engine BEGIN (implicit)
> 2024-03-13 11:52:45,134 INFO sqlalchemy.engine.Engine SAVEPOINT sa_savepoint_1
> 2024-03-13 11:52:45,134 INFO sqlalchemy.engine.Engine [no key 0.00024s] ()
> 2024-03-13 11:52:45,141 INFO sqlalchemy.engine.Engine INSERT INTO countries (code, name, is_active) VALUES ($1::VARCHAR, $2::VARCHAR, $3::BOOLEAN) RETURNING countries.id
> 2024-03-13 11:52:45,141 INFO sqlalchemy.engine.Engine [generated in 0.00028s] ('97276', '63910', True)
> 2024-03-13 11:52:45,147 INFO sqlalchemy.engine.Engine RELEASE SAVEPOINT sa_savepoint_1
> 2024-03-13 11:52:45,147 INFO sqlalchemy.engine.Engine [no key 0.00018s] ()
> 2024-03-13 11:52:45,155 INFO sqlalchemy.engine.Engine COMMIT
> 2024-03-13 11:52:45,179 INFO sqlalchemy.engine.Engine BEGIN (implicit)
> 2024-03-13 11:52:45,179 INFO sqlalchemy.engine.Engine INSERT INTO countries (code, name, is_active) VALUES ($1::VARCHAR, $2::VARCHAR, $3::BOOLEAN) RETURNING countries.id
> 2024-03-13 11:52:45,179 INFO sqlalchemy.engine.Engine [cached since 0.03847s ago] ('39570', '58035', True)
> 2024-03-13 11:52:45,187 INFO sqlalchemy.engine.Engine COMMIT

As you can see savepoint is being released after first commit.

**Reply by CaselIT** [reply]
I may not have understood what you want to do. You want to make a session use an external transaction? So this pattern? https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites

I don't think you can do that with a scoped session, since it's harder to control its lifetime.
Use a normal non-scoped session if you want to do this

**Reply by albertalexandrov** [reply]
> You want to make a session use an external transaction?

Yeap. 

The problem is that I cannot pass a different bind when instantiating session. The constraint is that session already exists. And I need to do with this session. I thought I could do something with the bind of the session to open an external transaction...

**Reply by CaselIT** [reply]
As mentioned a scoped session is not the right construct for what you want to so

**Reply by albertalexandrov** [reply]
Thanks for your time
